.github/workflows/ci.yml

```diff
@@ -9,30 +9,26 @@ on:
       - main
 
 jobs:
-  Tests:
+  nim-light-client-wasm-tests:
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v3
         with:
-          submodules: recursive
+          submodules: true
 
-      - uses: cachix/install-nix-action@v17
-        with:
-          nix_path: nixpkgs=channel:nixos-22.05
-          install_url: https://releases.nixos.org/nix/nix-2.8.1/install
-
-      - uses: cachix/cachix-action@v10
-        with:
-          name: nix-blockchain-development
+      - uses: ./.github/actions/build-nix-dev-shell
 
-      - name: Build Nix dev shell
-        run: nix build '.#devShell.x86_64-linux'
+      - name: Run Nim Beacon Chain Light Client WebAssembly tests
+        run: nix develop -c yarn test-emcc
 
-      - name: Install Node.js dependencies
-        run: nix develop -c yarn install --immutable
+  circom-tests:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          submodules: true
 
-      - name: Run emsctipten tests
-        run: nix develop -c yarn test-emcc
+      - uses: ./.github/actions/build-nix-dev-shell
```

Comment:
Doesn't this mean that Nix will have to install all dependencies in each job separately? This is unlikely to scale well in the long term.

Reply:
Here's a more detailed explanation of my reasoning.

Let's take an example scenario, hopefully, not too far from the real world:

* Let's say we have pipeline with 2 setup steps and 7 test steps. 
* The 2 setup steps need to be executed in order, before any of the 7 test steps start
* Any of the 7 steps can be run independently of the other 4
* I have labeled the time to execute each of this steps with the letters A..I
* Based on my experience, GH Actions is generally quite generous with number of concurrent runners that they provide. According to [this article](https://docs.github.com/en/actions/learn-github-actions/usage-limits-billing-and-administration#usage-limits) this number is 20 (but only 5 if they're all macOS based), but this sounds too good to be true :smile: 
  * Let's assume this number is 4 for the sake of discussion
  * That would mean, that if have `N` jobs (each taking the same amount of time for simplicity), the time to complete would be `N / 4`, since we would need to wait for the first 4 jobs to complete before any new jobs are started
  * Of course, in practice each job will take different time, and most likely a new runner would be started as soon as the first job of the first wave completes
  * In the table below I have listed 7 runners, but in practice, runners 5, 6 and 7 maybe the same as 1, 2 and 3. I have listed them as separate runners as they would still need to start the setup from the beginning and so it is easier to model them as separate runners which wait for first 4 to complete

| Step         | Single job / runner | Multiple jobs:<br>runner 1 | Multiple jobs:<br>runner 2 | Multiple jobs:<br>runner 3 | Multiple job:<br>runner 4 | Multiple job:<br>runner 5 | Multiple job:<br>runner 6 | Multiple job:<br>runner 7 |
|--------------|---------------------|----------------------------|----------------------------|----------------------------|---------------------------|---------------------------|---------------------------|---------------------------|
| Nix install  | A                   | A                          | A                          | A                          | A                         |                           |                           |                           |
| Yarn install | B                   | B                          | B                          | B                          | B                         |                           |                           |                           |
| Test 1       | C                   | C                          | D                          | E                          | F                         |                           |                           |                           |
| Test 2       | D                   |                            |                            |                            |                           | A                         | A                         | A                         |
| Test 3       | E                   |                            |                            |                            |                           | B                         | B                         | B                         |
| Test 4       | F                   |                            |                            |                            |                           | G                         | H                         | I                         |
| Test 5       | G                   |                            |                            |                            |                           |                           |                           |                           |
| Test 6       | H                   |                            |                            |                            |                           |                           |                           |                           |
| Test 7       | I                   |                            |                            |                            |                           |                           |                           |                           |

Total time to completion:
* Single job workflow: `sum(A, B, C, D, E, F, G, H, I)`
* Multi job workflow: `sum(A, B) + max(C, D, E, F) + sum(A, B) + max(G, H, I)`
  * If we have only 4 test jobs: `sum(A, B) + max(C, D, E, F)`
  * If we're given 7 parallel workers: `sum(A, B) + max(C, D, E, F, G, H, I)`

Reply:
The total number of runners in Github are limited by organisation though. We are suffering from this problem within the Status org where quite often build tasks spend a lot of time in the "queued" state.

Reply:
I believe what @PetarKirov proposed is a good practice. We use this in Hyperledger Cactus, although it bottlenecks eventually when the free tier is exhausted. 

Another advantage is that if one of the tests fails one can re-run a single and not the whole suite (which could be useful for non-deterministic tests, for example tests that pull docker containers from the web).